pkg/cmd/server/origin/auth.go

```diff
@@ -379,6 +382,16 @@ func (c *AuthConfig) getAuthenticationHandler(mux cmdutil.Mux, errorHandler hand
 			if identityProvider.UseAsChallenger {
 				return nil, errors.New("oauth identity providers cannot issue challenges")
 			}
+		} else if requestHeaderProvider, isRequestHeader := identityProvider.Provider.Object.(*configapi.RequestHeaderIdentityProvider); isRequestHeader {
```

Comment:
This method needs refactoring.  It wasn't pretty before, but now it's a 100 line monster that's nested pretty deeply.

I'm pretty sure that this does strange things if the config is combined with a password authenticator.